lib/IRGen/GenValueWitness.cpp

```diff
@@ -831,13 +832,22 @@ static void addValueWitness(IRGenModule &IGM,
 
   // Try to use a standard function.
   switch (index) {
-  case ValueWitness::Destroy:
+  case ValueWitness::Destroy: {
+    bool isClangType =
+        (concreteType.getStructOrBoundGenericStruct() &&
+         concreteType.getStructOrBoundGenericStruct()->getClangDecl());
+    if (isClangType) {
+      if (auto destructor = getCXXDestructor(concreteType))
+        return addFunction(getCXXDestructorCallee(IGM, destructor));
```

Comment:
Hm, I was wrong, `getCXXDestructorCallee` does get called a second time.

I'm not sure why this code path is necessary though? Why can't we just fall through to the `goto standard` and let the code you're adding in `ClangRecordTypeInfo::destroy` handle this?

Reply:
This code path is necessary. We need to make sure we store the correct function pointer (in this case the result of `getCXXDestructorCallee`) in the metadata/witness table. Otherwise, we'll do the wrong thing when we try to destroy a generic or existential type.

Reply:
`ClangRecordTypeInfo::destroy` only works when we know the type at compile time. With generics, we don't know what type we might get until runtime.

Reply:
But won't the `buildValueWitnessFunction` in the `goto standard` path package up the `ClangRecordTypeInfo::destroy` in an LLVM function and insert it into the VWT so we can then dynamically call it at runtime?

Reply:
PS Otherwise, how would this work for Swift types? For those, we only special-case for `isPOD` and `isSingleSwiftRetainablePointer`. Everything else relies on the `TypeInfo`.

Reply:
Yes, you're right. That works.